### PR TITLE
[WIP] Capsnet layers

### PR DESCRIPTION
**Pull request by rnett**
## What changes were proposed in this pull request?
From #7386

Adds DigiCaps, PrimaryCaps, and a capsule strength layer from Dynamic Routing Between Capsules, as CapsuleLayer, PrimaryCapsules, and CapsuleStrengthLayer, respectively, as well as a utility class.

All are implemented using SameDiff.

## How was this patch tested?

None yet.  Will need tests for all 3 layers, as well as bad configuration tests.

## TODO
- [x] CapsuleLayer
- [x] PrimaryCapsules
- [x] CapsuleStrengthLayer
- [x] Docstrings
- [x] Tests
- [ ] Arbiter?

### CODE REVIEW

**Comment by rnett** [comment]
Planed tests:
- [ ] CapsuleLayer
  - [ ] outputType
  - [ ] input type inference
  - [ ] config
  - [ ] output shape of single layer network
- [ ] PrimaryCapsules
  - [ ] outputType
  - [ ] input type inference
  - [ ] config
  - [ ] output shape of single layer network
- [ ] CapsuleStrength
  - [ ] outputType
  - [ ] config
  - [ ] output shape of single layer network
- [ ] MNIST 5 epoch?

**Comment by treo** [comment]
Maybe also add gradient checks for your layers here. As it is still early days for samediff, a more sophisticated use of it, like your layers, may surface some bugs. 

Gradient checks for layers are pretty simple to do, for an example see https://github.com/deeplearning4j/deeplearning4j/pull/7311/files#diff-956c3d9bc48af5c418208bce04cd0027

**Comment by rnett** [comment]
I will do that.  I ran into a couple of bugs and reported them, but what I have currently seems to be working, I can get >95% accuracy on MNIST with 2 epochs and a very scaled-down network compared to the original CapsNet.  Of course, its possible there is a issue somewhere in there that just isn't big enough to be detected, but it looks fairly good.

**Comment by rnett** [comment]
I'm getting errors in the gradient check.  However, the reported gradients are relatively close, so I'm not sure if its my code or compounding SameDiff errors.

Report:
```
nParams, layer 0: 1312
nParams, layer 1: 50240
nParams, layer 2: 10240
nParams, layer 3: 0
minibatch=64, PrimaryCaps: 8 channels, 8 dimensions, Capsules: 10 capsules with 16 dimensions and 3 routings
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Layer 0: ConvolutionLayer - params [b, W]
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Layer 1: SameDiffLayer - params [weight, bias]          // PrimaryCapsules, these are parameters for the convolution
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Layer 2: SameDiffLayer - params [weight]                // CapsuleLayer
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Layer 3: SameDiffLayer - params []                      // CapsuleStrengthLayer
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Layer 4: ActivationLayer - params []
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Layer 5: LossLayer - params []
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 0 (0_b) FAILED: grad= -2.379222777046915E-4, numericalGrad= -2.300193369109138E-4, relError= 0.016888732583165585, scorePlus=2.3062806561712503, scoreMinus= 2.306280656631289, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 1 (0_b) passed: grad= 5.929871790586867E-4, numericalGrad= 5.91871440747127E-4, relError= 9.416636659507323E-4
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 2 (0_b) FAILED: grad= -3.268174111421275E-4, numericalGrad= -3.2263214322370004E-4, relError= 0.006444331034323755, scorePlus=2.3062806560786377, scoreMinus= 2.306280656723902, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 3 (0_b) FAILED: grad= 1.9390171313378663E-4, numericalGrad= 1.9152812669176456E-4, relError= 0.006158284068240222, scorePlus=2.306280656592798, scoreMinus= 2.3062806562097418, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 4 (0_b) FAILED: grad= 4.883863696451556E-4, numericalGrad= 4.786830931635677E-4, relError= 0.01003369132699733, scorePlus=2.306280656879953, scoreMinus= 2.306280655922587, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 5 (0_b) FAILED: grad= -5.9985731824785914E-5, numericalGrad= -6.629985449535525E-5, relError= 0.049998759593693244, scorePlus=2.3062806563349705, scoreMinus= 2.30628065646757, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 6 (0_b) FAILED: grad= 2.681832847767592E-4, numericalGrad= 2.6690916143934373E-4, relError= 0.0023811274975481696, scorePlus=2.3062806566681795, scoreMinus= 2.306280656134361, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 7 (0_b) FAILED: grad= 7.39580096531447E-4, numericalGrad= 7.302829452271453E-4, relError= 0.006325182034088197, scorePlus=2.306280657131553, scoreMinus= 2.306280655670987, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 8 (0_b) FAILED: grad= -6.130691658394211E-6, numericalGrad= -8.705702825295702E-6, relError= 0.17356044082895455, scorePlus=2.3062806563925644, scoreMinus= 2.306280656409976, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 9 (0_b) FAILED: grad= 4.509005322153342E-4, numericalGrad= 4.4923442743538544E-4, relError= 0.0018509499737631159, scorePlus=2.3062806568505048, scoreMinus= 2.306280655952036, paramValue = 0.0
[main] INFO org.deeplearning4j.gradientcheck.GradientCheckUtil - Param 10 (0_b) FAILED: grad= 1.9728276810339123E-4, numericalGrad= 1.9635360004599534E-4, relError= 0.0023604730979615053, scorePlus=2.3062806565976235, scoreMinus= 2.3062806562049163, paramValue = 0.0
```

**Comment by rnett** [comment]
Alright, so the tests pass if you set routings to 1.  Routings is the loop limit for the internal SameDiff loop, so either this is a SameDiff compounding errors issue, or there is something in the [post-short circut operations](https://github.com/deeplearning4j/deeplearning4j/blob/4a34a454ae5a5118a14c87da373047044a9fe731/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/CapsuleLayer.java#L123-L133) that causes this.

Tests should now all be passing, I'm just adding a few comments.

**Comment by treo** [comment]
Make sure that you've got the latest master merged into this and run the gradient checks again. There was an actual bug in samdiff were recurrent gradients under some circumstances were not added into the total. https://github.com/deeplearning4j/deeplearning4j/pull/7393 you should have these changes on your branch for the fix

**Comment by rnett** [comment]
It is in there.  No change in the test.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/util/CapsuleUtils.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     *  Canceling one ||s|| gives ||s||*s/((1 + ||s||^2)
+     *
+     * @param SD The SameDiff environment
+     * @param x The variable to squash
+     * @return squash(x)
+     */
+    public static SDVariable squash(SameDiff SD, SDVariable x, int dim){
+        SDVariable squaredNorm = SD.math.square(x).sum(true, dim);
+        SDVariable scale = SD.math.sqrt(squaredNorm.plus(1e-5));
+        return x.times(squaredNorm).div(squaredNorm.plus(1.0).times(scale));
```
[the remark this answers is not in the document]
Comment (rnett) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/util/CapsuleUtils.java]: That was because I saw it in every implementation I looked at and wasn't sure why, and so didn't want to remove it.

From [here](https://github.com/tensorflow/tensorflow/issues/4914), it seems to be to keep the gradient stable if the squaredNorm is zero.  Is that an issue SameDiff will have?

**Comment by rnett** [comment]
For issues 1 and 3, part of the reason I did it that way is I had no way to get the minibatch size, because of variable sized minibatches.  Is there a way around that?  E.g. for 3, I couldn't reshape because I don't know what the minibatch size would be, and I can't use -1 there.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/util/CapsuleUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexDBlack) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/util/CapsuleUtils.java]: Ah, makes sense, let's leave it then.

**Comment by AlexDBlack** [comment]
Re: no minibatch size - it's possible, but on reflection it's not easy... OK, let's leave it for now until we have something better: https://github.com/deeplearning4j/deeplearning4j/issues/7445

As for the `SD.zero("b_var", 1, inputCapsules, capsules, 1, 1);` - that'll make it a broadcast softmax (same value for all examples in minibatch) later in first iteration `SDVariable c = CapsuleUtils.softmax(SD, b, 2, 5);` which is not what we want... Maybe let's change this back, then this should be good to go (we'll make a note on that issue I just opened for the inefficient zeros/subset thing, and fix it in same PR)
